qemu/tests/slof_user_interface.py

```diff
@@ -0,0 +1,178 @@
+      2.4. Check guest whether boot up successfully.
+      2.5. Log in guest successfully.
+
+    :param test: Qemu test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+    def _create_iso():
+        """Create iso with one file on it."""
+        if params.get("create_cd", 'no') == 'yes':
+            logging.info("creating iso.")
+            iso_name = params.get("cdrom_cd1")
+            iso = utils_misc.get_path(data_dir.get_data_dir(), iso_name)
+            process.system_output("dd if=/dev/urandom of=new bs=10M count=1",
+                                  300)
+            process.system_output("mkisofs -o %s new" % iso, 300)
+            process.system_output("rm -f new", 300)
+
+    def _cleanup_iso():
+        """Removes created cdrom."""
+        if params.get("create_cd", 'no') == 'yes':
+            logging.info("cleaning up temp iso")
+            cdrom_test = utils_misc.get_path(
+                data_dir.get_data_dir(), params.get("cdrom_cd1"))
+            os.remove(cdrom_test)
+
+    def _send_custom_key(keystr):
+        """Send custom keyword to SLOF's user interface."""
+        logging.info('Sending \"%s\" to SLOF user interface.' % keystr)
+        for key in keystr:
+            if key == '-':
+                vm.send_key('minus')
+            else:
+                vm.send_key(key)
+        vm.send_key('ret')
+
+    def _send_key(key, custom=False, sleep=0.0):
+        """Send keywords to SLOF's user interface."""
+        obj_name = 'select' if re.search(r'^\d+$', key) else key
+        key_params = params.object_params(obj_name)
+        if not custom:
+            vm.send_key(key)
+        else:
+            _send_custom_key(key)
+        time.sleep(sleep)
+        content, _ = slof.get_boot_content(vm,
+                                           start_str=key_params.get('start'),
+                                           end_str=key_params.get('end'))
+        if content:
+            logging.info('Output of SLOF:\n%s' % ''.join(content))
+            return ''.join(content)
+        return None
+
+    def _check_menu_info(menu_info):
+        """Check the menu info by each items."""
+        bootable_num = ''
+        for i in range(1, int(params['boot_dev_num']) + 1):
+            option = params['menu_option%d' % i]
+            logging.info('Checking the device(%s) if is included in menu list.'
+                         % '->'.join(option.split()))
+
+            dev_type, hba_type, child_bus, addr = tuple(option.split())
+            addr = re.sub(r'^0x0?', '', addr)
+            pattern = re.compile(r'(\d+)\)\s+%s(\d+)?\s+:\s+/%s(\S+)?/%s@%s' %
+                                 (dev_type, hba_type, child_bus, addr), re.M)
+            searched = pattern.search(menu_info)
+            if not searched:
+                test.fail('No such item(%s) in boot menu list.' %
+                          '->'.join(option.split()))
```

Comment:
if pass, it's better to give related pass log.

Reply:
As talked with xianwang, cancel this requirement.